File: sem1/hw11/hw-11.3/hw-11.3/PriorityQueue.cpp
#include "PriorityQueue.h"

using namespace std;

struct QueueElement
{
	int key = 0;
	int node = 0;
	int	adjacentNode = 0;
	QueueElement *next;
};

struct PriorityQueue
{
	QueueElement *head = nullptr;
};

PriorityQueue *createPriorityQueue()
{
	return new PriorityQueue;
}

bool isEmpty(PriorityQueue *queue)
{
	return queue->head == nullptr;
}

void enqueue(PriorityQueue *queue, const int node, const int adjacentNode, const int key)
{
	QueueElement *newElement = new QueueElement{ key, node, adjacentNode, nullptr };
	if (isEmpty(queue))
	{
		queue->head = newElement;
		return;
	}

	if (queue->head->key >= newElement->key)
	{
		newElement->next = queue->head;
		queue->head = newElement;
		return;
	}

	QueueElement *previous = queue->head;
	QueueElement *current = queue->head->next;
	while (current != nullptr && current->key < newElement->key)
	{
		previous = current;
		current = current->next;
	}
	newElement->next = current;
	previous->next = newElement;
}

int dequeue(PriorityQueue *queue, int &node)
{
	QueueElement *temp = queue->head;
	queue->head = queue->head->next;
	int adjacentNode = temp->adjacentNode;
	node = temp->node;
	delete temp;
	return adjacentNode;
}

void deleteQueue(PriorityQueue *queue)
{
	while (!isEmpty(queue))
	{
		QueueElement *temp = queue->head;
		queue->head = queue->head->next;
		delete temp;
	}
	delete queue;
}

File: sem2/tests/test1/test1/PriorityQueue.cs
using System;

namespace test1
{
    /// <summary>
    /// Queue data structure where each element has a priority associated with it.
    /// </summary>
    public class PriorityQueue
    {
        private class Element
        {
            public string Data { get; set; }
            public int Priority { get; set; }
            public Element Next;

            public Element(string data, int priority)
            {
                Data = data;
                Priority = priority;
            }
        }

        private Element head;

        /// <summary>
        /// Checks if the priority queue has no elements in it.
        /// </summary>
        /// <returns>True if the priority queue is empty, else false.</returns>
        public bool IsEmpty() => head == null;

        /// <summary>
        /// Adds new queue element in the correct place depending on the given priority.
        /// </summary>
        /// <param name="data">Element to add.</param>
        /// <param name="priority">Priority of the element.</param>
        public void Enqueue(string data, int priority)
        {
            var newElement = new Element(data, priority);
            
            if (IsEmpty())
            {
                head = newElement;
                return;
            }

            if (head.Priority < priority)
            {
                newElement.Next = head;
                head = newElement;
                return;
            }

            Element previous = head;
            Element current = head.Next;

            while (current != null && current.Priority >= priority)
            {
                previous = current;
                current = current.Next;
            }

            newElement.Next = current;
            previous.Next = newElement;
        }


        /// <summary>
        /// Returns element with the highest priority and removes it from the queue.
        /// </summary>
        /// <returns>Queue element with the highest priority</returns>
        public string Dequeue()
        {
            if (IsEmpty())
            {
                throw new QueueIsEmptyException();
            }

            Element temp = head;
            head = head.Next;

            return temp.Data;
        }

        /// <summary>
        /// Prints the queue шn descending order of priority.
        /// </summary>
        public void PrintQueue()
        {
            Element current = head;

            while (current != null)
            {
                Console.Write($"{current.Data} ");
                current = current.Next;
            }
            Console.WriteLine();
        }
    }
}
